packages/matter.js/src/storage/StorageContext.ts

```diff
@@ -4,57 +4,96 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { Storage, StorageError } from "./Storage.js";
+import { MaybePromise } from "../util/Promises.js";
+import { Storage, StorageError, SyncStorage } from "./Storage.js";
 import { SupportedStorageTypes } from "./StringifyTools.js";
 
-export class StorageContext {
+export class StorageContext<S extends Storage = any> {
     constructor(
-        private readonly storage: Storage,
-        readonly contexts: string[],
+        private readonly storage: S,
+        readonly thisContexts: string[],
     ) {}
 
-    get<T extends SupportedStorageTypes>(key: string, defaultValue?: T): T {
-        const value = this.storage.get<T>(this.contexts, key);
-        if (value !== undefined) return value;
+    get<T extends SupportedStorageTypes>(key: string, defaultValue?: T): S extends SyncStorage ? T : Promise<T> {
+        const value = this.storage.get(this.thisContexts, key);
+        if (value !== undefined) {
+            if (MaybePromise.is(value)) {
+                return value.then(v => {
+                    if (v !== undefined) return v;
+                    if (defaultValue === undefined) {
+                        throw new StorageError(
+                            `No value found for key ${key} in context ${this.thisContexts} and no default value specified!`,
+                        );
+                    }
+                    return defaultValue;
+                }) as S extends SyncStorage ? T : Promise<T>;
+            }
+            return value as S extends SyncStorage ? T : Promise<T>;
```

Comment:
See my comment above r.e. StorageOperationResult utility type, I think it'd be an improvement, although maybe could shorten to "StorageResult" since it'll be used more than I was thinking..